File: algo/424.py
from collections import Counter

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        counts = Counter()
        left = 0
        for right, v in enumerate(s, 1):
            counts[v] += 1
            max_char_n = counts.most_common(1)[0][1]
            if right - left - max_char_n > k:
                counts[s[left]] -= 1
                left += 1
        return right - left

s = "AAABBC"
k = 2
print(Solution().characterReplacement(s, k))

File: old/algo/424.kt
class Solution {
    fun characterReplacement(s: String, k: Int): Int {
        var left = 0
        val counts = mutableMapOf<Char, Int>()
        var start = 0
        var end = 0
        for ((right, ch) in s.withIndex()) {
            counts[ch] = counts.getOrDefault(ch, 0) + 1
            var maxCharLen = counts.entries.maxBy { it.value }.value
            var target = right - left + 1 - maxCharLen

            if (target > k) {
                val leftChar = s[left]
                counts[leftChar] = counts.getValue(leftChar) - 1
                left++
            }
            start = left
            end = right
        }
        return end - start + 1
    }
}

//val s = "ABAB"
//val k = 2

//val s = "AABABBA"
//val k = 1

val s = "AAAA"
val k = 2

println(Solution().characterReplacement(s, k))
